### PR TITLE
Dropbox/reports

### PR DESCRIPTION
**Pull request by ValbuenaG**
For more information on contributing to Tella source code, see the [Contributor Guidelines](contributing/contributor_guide.md).

## Type of change

**Description:**

**Select the type of change(s) made in this pull request:**
- [ ] Bug fix *(Fixes an issue)*
- [ ] New feature *(Adds functionality)*
- [ ] Documentation *(Fix to documentation)*

----------------------------------------------------------------------------------------

Fixes #issue-number

## Proposed changes 

*
*
*

### CODE REVIEW

**Review by Rimktarii — CHANGES_REQUESTED** [review]

File: Tella/Domain/Entity/Dropbox/DropboxReportFile.swift [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+//  Tella
+//
+//  Created by gus valbuena on 9/26/24.
+//  Copyright © 2024 HORIZONTAL. All rights reserved.
+//
+
+import Foundation
+
+class DropboxReportFile: ReportFile {
+    var offset: Int64?
```
Comment (dhekra-rouatbi) [inline Tella/Domain/Entity/Dropbox/DropboxReportFile.swift:12]: Can we use bytesSent ?

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Data/Networking/APICall/APIError.swift [hunk 2]
```diff
@@ -39,6 +41,8 @@ extension APIError: LocalizedError {
             return customNcErrorMessage(errorCode: code)
         case .errorOccured :
             return LocalizableError.commonError.localized
+        case .dropboxApiError(let error):
+            return DropboxErrorParser.customDropboxErrorMessage(error: error)
```
Comment (dhekra-rouatbi) [inline Tella/Data/Networking/APICall/APIError.swift:45]: What do you think of using an `APIError`  extension that contains `customDropboxErrorMessage` function ?

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Scenes/Dropbox/ViewModel/DropboxOutboxViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            do {
+                try await dropboxRepository.handleSignIn()
+            } catch {
+                debugLog(error)
+            }
+        }
+    }
+    
+    func handleURLRedirect(url: URL) {
+        _ = dropboxRepository.handleRedirectURL(url) { authResult in
```
Comment (dhekra-rouatbi) [inline Tella/Scenes/Dropbox/ViewModel/DropboxOutboxViewModel.swift:204]: maybe we need to handle success and failure ? like in `DropboxServerViewModel` class

    func handleURLRedirect(url: URL) {
        _ = dropboxRepository.handleRedirectURL(url) { [weak self] authResult in
            guard let self = self else { return }
            
            switch authResult {
            case .success:
                self.signInState = .loaded(true)
                self.addServer()
            case .error(_, let description):
                self.signInState = .error(description ?? "")
            default:
                break
            }
        }
    }

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Domain/Entity/Report/ReportVaultFile.swift [hunk 4]
```diff
@@ -17,6 +17,8 @@ class ReportVaultFile : VaultFileDB {
     var reportInstanceId : Int?
     var chunkFiles: [(fileName: String, size: Int64)]?
     var finishUploading : Bool = false
+    var offset: Int64?
```
Comment (dhekra-rouatbi) [inline Tella/Domain/Entity/Report/ReportVaultFile.swift:20]: maybe here we can use `bytesSent` instead of adding an other `offset` variable

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Domain/Entity/Dropbox/DropboxReportFile.swift [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    
+    init(id: Int? = nil,
+         fileId: String? = nil,
+         status: FileStatus? = nil,
+         bytesSent: Int? = 0,
+         createdDate: Date? = nil,
+         updatedDate: Date? = Date(),
+         reportInstanceId: Int? = nil,
+         offset: Int64? = nil,
```
Comment (dhekra-rouatbi) [inline Tella/Domain/Entity/Dropbox/DropboxReportFile.swift:25]: I think `offset` is unused, could you remove it

File: Tella/Data/Database/Common/DatabaseConstants.swift [hunk 6]
```diff
@@ -109,6 +110,10 @@ struct D {
     static let cUserId = "c_user_id"
     static let cRemoteReportStatus = "c_remote_report_status"
     static let cChunkFiles = "c_chunk_files"
+    
+    //dropbox
+    static let cOffset = "c_offset"
```
Comment (dhekra-rouatbi) [inline Tella/Data/Database/Common/DatabaseConstants.swift:115]: I think `cOffset ` is unused, could you remove it

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Data/Networking/Repositories/Dropbox/DropboxErrorParser.swift [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+//  Tella
+//
+//  Created by gus valbuena on 10/2/24.
+//  Copyright © 2024 HORIZONTAL. All rights reserved.
+//
+
+import Foundation
+import SwiftyDropbox
+
+extension APIError {
```
Comment (dhekra-rouatbi) [inline Tella/Data/Networking/Repositories/Dropbox/DropboxErrorParser.swift:12]: maybe we don't need a `static` for  `customDropboxErrorMessage` function ?

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Scenes/Dropbox/ViewModel/DropboxOutboxViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    }
+    
+    
+    override func submitReport() {
+        if isSubmissionInProgress { return }
+
+        self.updateReportStatus(reportStatus: .submissionInProgress)
+        cancellables.removeAll()
+
+        if reportViewModel.folderId != nil {
```
Comment (dhekra-rouatbi) [inline Tella/Scenes/Dropbox/ViewModel/DropboxOutboxViewModel.swift:59]: I think it's better to move the upload handling to the `DropboxRepository` and handling the result in the view model. 
There is an example  in `NextcloudOutboxViewModel` : `performSubmission()` function

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+                fileURL: file.url,
+                fileName: file.fileName,
+                fileId: file.fileId,
+                sessionId: file.sessionId,
+                offset: file.offset ?? 0,
+                totalBytes: (try? FileManager.default.attributesOfItem(atPath: file.url.path)[.size] as? Int64) ?? 0
+            )
+            
+            do {
+                try await uploadFile(
```
Comment (dhekra-rouatbi) [inline Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift:214]: In my opinion Sending all the files in parallel is better than the serial upload to follow Tella web and Nextcloud implementation

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+        try await ensureSignedIn()
+        return try await performCreateFolder(name: name, description: description)
+    }
+    
+    private func performCreateFolder(name: String, description: String) async throws -> DropboxCreateFolderResponse {
+        guard let client = self.client else {
+            throw APIError.noToken
+        }
+        
+        let folderPath = "/\(name)"
```
Comment (dhekra-rouatbi) [inline Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift:135]: can we create an extension to add "/" as a prefix ?

there is an extension to add a "/" as a prefix 
```
    func slash() -> String {
        return self.hasSuffix("/") ? self : self + "/"
    }

```

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

File: Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+                if let folderId = folderId {
+                    folderResponse = DropboxCreateFolderResponse(id: folderId, name: name)
+                } else {
+                    folderResponse = try await createFolder(name: name, description: description)
+                    subject.send(.folderCreated(folderId: folderResponse.id, folderName: folderResponse.name))
+                }
+                
+                // Upload files
+                try await uploadFiles(
+                    to: "/\(folderResponse.name)",
```
Comment (dhekra-rouatbi) [inline Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift:106]: https://github.com/Horizontal-org/Tella-iOS/pull/189#discussion_r1793812638

File: Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift [hunk 12]
```diff
@@ -0,0 +1,465 @@
...
+                    continuation.resume(throwing: APIError.errorOccured)
+                }
+            }
+        }
+        
+        let folderId = metadata.id
+        let folderName = metadata.name
+        
+        // Upload the description file if it exists
+        try await uploadDescriptionFile(client: client, folderPath: "/\(folderName)", description: description)
```
Comment (dhekra-rouatbi) [inline Tella/Data/Networking/Repositories/Dropbox/DropboxRepository.swift:157]: https://github.com/Horizontal-org/Tella-iOS/pull/189#discussion_r1793812638

**Review by dhekra-rouatbi — CHANGES_REQUESTED** [review]

**Pull request merged**